libnd4j/include/ops/declarable/helpers/cpu/random.cpp

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright (c) 2015-2018 Skymind, Inc.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+//
+//  @author sgazeos@gmail.com
+//
+
+#include <ops/declarable/helpers/random.h>
+//#include <vector>
+#include <memory>
+//#include <graph/Context.h>
+#include <ShapeUtils.h>
+
+namespace nd4j {
+namespace ops {
+namespace helpers {
+
+    template <typename T>
+    void fillRandomGamma_(LaunchContext* context, graph::RandomGenerator& rng, NDArray* alpha, NDArray* beta, NDArray* output) {
+
+        Nd4jLong* broadcasted = nullptr;
+        if (beta != nullptr)
+            ShapeUtils::evalBroadcastShapeInfo(*alpha, *beta, true, broadcasted, context->getWorkspace());
+        else
+            broadcasted = alpha->shapeInfo();
+        auto step = shape::length(broadcasted);
+        auto shift = output->lengthOf() / step;
+
+        auto copyAlpha = alpha;
+        auto copyBeta = beta;
+        if (beta != nullptr) {
+            NDArray alphaBroadcasted(broadcasted, alpha->dataType(), false, context);
+            NDArray betaBroadcasted(broadcasted, beta->dataType(), false, context);
+
+            copyAlpha = (alphaBroadcasted.applyTrueBroadcast(BroadcastOpsTuple::Assign(), alpha));
+            copyBeta = (betaBroadcasted.applyTrueBroadcast(BroadcastOpsTuple::Assign(), beta));
+
+        }
+
+        PRAGMA_OMP_PARALLEL_FOR
+        for (auto k = 0; k < shift; k++) {
+            auto pos = k * step;
+            auto u = rng.relativeT<T>(k, 0., 1.);
+            for (auto e = 0; e < step; e++)
+                    output->t<T>(pos + e) = math::nd4j_igamma<T, T, T>(copyAlpha->t<T>(e),
```

Comment:
Make it work on buffer please, instead of NDArray methods. Pretty much like you did it for CUDA